### PR TITLE
Improve `get_context_object_name` on `SingleObjectMixin` and `MultipleObjectMixin`

### PR DESCRIPTION
**Pull request by flaeppe**
Get in an overload for method `get_context_object_name`

## Related issues

- Refs #873

### CODE REVIEW

File: django-stubs/views/generic/detail.pyi [hunk 1]
```diff
@@ -7,7 +7,7 @@ from django.views.generic.base import ContextMixin, TemplateResponseMixin, View
 _M = TypeVar("_M", bound=models.Model)
 
 class SingleObjectMixin(Generic[_M], ContextMixin):
-    model: type[_M]
+    model: type[_M] | None
```
Comment (sobolevn) [inline django-stubs/views/generic/detail.pyi]: This can be a breaking change for a lot of code.

Comment (flaeppe) [reply django-stubs/views/generic/detail.pyi]: Yeah perhaps. Its default value is `None` though. So if you're declaring a view, subclassing `SingleObjectMixin` and also declare `model`. The type will be that model, `None` would go away.

Do you think we should add that as a test?

Any other, more generic, code expecting this kind of view should hopefully be a bit happy that `None` is covered.

File: django-stubs/views/generic/list.pyi [hunk 2]
```diff
@@ -8,7 +8,7 @@ from django.views.generic.base import ContextMixin, TemplateResponseMixin, View
 
 _M = TypeVar("_M", bound=Model)
 
-class HasModel(Protocol):
+class _HasModel(Protocol):
```
Comment (sobolevn) [inline django-stubs/views/generic/list.pyi:12]:
```suggestion
@type_check_only
class _HasModel(Protocol):
```

File: django-stubs/views/generic/detail.pyi (hunk 1 above)
Comment (sobolevn) [reply django-stubs/views/generic/detail.pyi]: Yes, sounds good.

Comment (flaeppe) [reply django-stubs/views/generic/detail.pyi]: > [...] So if you're declaring a view, subclassing SingleObjectMixin and also declare model. The type will be that model, None would go away. [...]

Well, so I thought, `pyright` behaves differently in regards to this and keeps the original type 🙂

Comment (sobolevn) [reply django-stubs/views/generic/detail.pyi]: I think that we have to decide: practicality or purity. I think that the first one is more important in this case. `None` is just a default / sentinel value.

Comment (flaeppe) [reply django-stubs/views/generic/detail.pyi]: Sure, I don't have that strong opinions about the type here but I would consider it fairly easy to run in to an `AttributeError: 'NoneType' object has no attribute 'model'`

Comment (flaeppe) [reply django-stubs/views/generic/detail.pyi]: But if it was a larger issue, someone would've probably reported it by now. So it should probably work out just fine

**Review by sobolevn — APPROVED** [review]

**Pull request merged**